src/glib/mine/avltree.cpp

```diff
@@ -0,0 +1,259 @@
+            node = parent;
+            parent = GetParentNode(parent);
+        }
+        return parent;
+    }
+}
+
+bool AvlTree::Add(TFlt x, TInt w) {
+	//printf("root: %f\n", Root);
+    if(Root == 0) {
+        Root = ++N;
+        Values.AddDat(Root, x);
+        Count.AddDat(Root, w);
+        Left.AddDat(Root, 0);
+        Right.AddDat(Root, 0);
+        Parent.AddDat(Root,0);
+        // Update depth and aggregates
+        UpdateAggregates(Root);
+        return true;
+    } else {
+        TInt node = Root;
+        TInt parent = 0;
+        TInt cmp;
+        do {
+            cmp = Compare(node, x);
+            if(cmp < 0) {
+                parent = node;
+                node = GetLeftNode(node);
+            } else if (cmp > 0) {
+                parent = node;
+                node = GetRightNode(node);
+            } else {
+                // we merge the node
+                Merge(node, x, w);
+                return false;
+            }
+        } while(node != 0);
+
+        node = ++N;
+        Values.AddDat(node, x);
+        Count.AddDat(node, w);
+        Left.AddDat(node, 0);
+        Right.AddDat(node, 0);
+        Parent.AddDat(node, parent);
+        if(cmp < 0) {
+            Left.AddDat(parent, node);
+        } else {
+            //assert(cmp > 0);
+            Right.AddDat(parent, node);
+        }
+
+        Rebalance(node);
+
+        return true;
+    }
+}
+
+TInt AvlTree::Find(TFlt x) const {
+    for(TInt node = Root; node != 0;) {
+        const TInt cmp = Compare(node, x);
+        if(cmp < 0) {
+            node = GetLeftNode(node);
+        } else if(cmp > 0) {
+            node = GetRightNode(node);
+        } else {
+            return node;
+        }
+    }
+    return 0;
+}
+
+TInt AvlTree::Floor(TFlt x) const {
+    TInt f = 0;
+    for(TInt node = Root; node != 0; ) {
+        const TInt cmp = Compare(node, x);
+        if(cmp <= 0) {
+            node = GetLeftNode(node);
+        } else {
+            f = node;
+            node = GetRightNode(node);
+        }
+    }
+    return f;
+}
+
+TInt AvlTree::FloorSum(long sum) const {
+    TInt f = 0;
+    for(TInt node = Root; node != 0; ) {
+        const TInt left = GetLeftNode(node);
+        const long leftCount = GetAggregatedCount(left);
+        if(leftCount <= sum) {
+            f = node;
+            sum -= leftCount + GetCount(node);
+            node = GetRightNode(node);
+        } else {
+            node = GetLeftNode(node);
+        }
+    }
+    return f;
+}
+
+long AvlTree::CeilSum(TInt node) const {
+    const TInt left = GetLeftNode(node);
+    long sum = GetAggregatedCount(left);
+    TInt n = node;
+    for(TInt p = GetParentNode(node); p != 0; p = GetParentNode(n)) {
+        if(n == GetRightNode(p)) {
+            const TInt leftP = GetLeftNode(p);
+            sum += GetCount(p) + GetAggregatedCount(leftP);
+        }
+        n = p;
+    }
+    return sum;
+}
+
+void AvlTree::Rebalance(TInt node) {
```

Comment:
const int& NodeN
Use TInt for: member variables (so they have default initialized values and can be serialized,...)